components/util/str.rs

```diff
@@ -159,72 +160,73 @@ fn is_ascii_digit(c: &char) -> bool {
 }
 
-fn is_decimal_point(c: char) -> bool {
-    c == '.'
+fn is_decimal_point(c: &char) -> bool {
+    *c == '.'
 }
 
-fn is_exponent_char(c: char) -> bool {
-    match c {
+fn is_exponent_char(c: &char) -> bool {
+    match *c {
         'e' | 'E' => true,
         _ => false,
     }
 }
 
-pub fn read_numbers<I: Iterator<Item=char>>(mut iter: Peekable<I>) -> (Option<i64>, usize) {
+pub fn read_numbers<I>(iter: &mut Peekable<I>) -> Option<i64>
+    where I: Iterator<Item=char> + Clone {
     match iter.peek() {
         Some(c) if is_ascii_digit(c) => (),
-        _ => return (None, 0),
-    }
+        _ => return None,
+    };
 
-    iter.take_while(is_ascii_digit).map(|d| {
+    iter.take_while_ref(is_ascii_digit).map(|d| {
         d as i64 - '0' as i64
-    }).fold((Some(0i64), 0), |accumulator, d| {
-        let digits = accumulator.0.and_then(|accumulator| {
+    }).fold(Some(0i64), |accumulator, d| {
+        accumulator.and_then(|accumulator| {
             accumulator.checked_mul(10)
         }).and_then(|accumulator| {
             accumulator.checked_add(d)
-        });
-        (digits, accumulator.1 + 1)
+        })
     })
 }
 
-pub fn read_fraction<I: Iterator<Item=char>>(mut iter: Peekable<I>,
-                                             mut divisor: f64,
-                                             value: f64) -> (f64, usize) {
+pub fn read_fraction<I>(iter: &mut Peekable<I>, mut divisor: f64, value: &mut f64)
+    where I: Iterator<Item=char> + Clone {
     match iter.peek() {
-        Some(c) if is_decimal_point(*c) => (),
-        _ => return (value, 0),
-    }
+        Some(c) if is_decimal_point(c) => (),
+        _ => return,
+    };
     iter.next();
 
-    iter.take_while(is_ascii_digit).map(|d|
+    *value = iter.take_while_ref(is_ascii_digit).map(|d|
         d as i64 - '0' as i64
-    ).fold((value, 1), |accumulator, d| {
+    ).fold(*value, |accumulator, d| {
         divisor *= 10f64;
-        (accumulator.0 + d as f64 / divisor,
-            accumulator.1 + 1)
-    })
+        accumulator + (d as f64) / divisor
+    });
 }
 
-pub fn read_exponent<I: Iterator<Item=char>>(mut iter: Peekable<I>) -> Option<i32> {
+pub fn read_exponent<I>(iter: &mut Peekable<I>) -> Option<i32>
+    where I: Iterator<Item=char> + Clone {
+
     match iter.peek() {
-        Some(c) if is_exponent_char(*c) => (),
+        Some(c) if is_exponent_char(c) => (),
         _ => return None,
-    }
+    };
     iter.next();
 
-    match iter.peek() {
-        None => None,
+    let exp_sign = match iter.peek() {
         Some(&'-') => {
             iter.next();
-            read_numbers(iter).0.map(|exp| -exp.to_i32().unwrap_or(0))
-        }
+            -1
+        },
         Some(&'+') => {
             iter.next();
-            read_numbers(iter).0.map(|exp| exp.to_i32().unwrap_or(0))
-        }
-        Some(_) => read_numbers(iter).0.map(|exp| exp.to_i32().unwrap_or(0))
-    }
+            1
+        },
+        Some(_) | None => 1,
```

Comment:
You've changed the behaviour of `None` with your refactoring. Show me the spec that backs up this change.